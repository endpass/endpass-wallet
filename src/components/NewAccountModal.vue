<template lang="html">
  <div class="new-account-modal">
    <v-modal @close="close">
      <template slot="header">Create New Address</template>

      <div v-if="!createdAccount">
        <p class="subtitle">You currently have
        <strong>{{accounts.length}}</strong> active addresses in your wallet.</p>
        <p class="subtitle">Click the button below to create an additional
      address you can use to receive Ethereum and tokens.</p>
      </div>
      <div v-else>
        <p class="subtitle">New Address Created</p>

        <div class="message">
          <div class="message-header">
            <p>Public Address</p>
          </div>
          <div class="message-body">
            <p>Use this address to receive Ether and tokens.</p>
            <p class="code address">{{createdAccount.getAddressString()}}</p>
          </div>
        </div>

        <div class="message is-warning">
          <div class="message-header">
            <p>Private Key</p>
          </div>
          <div class="message-body">
            <p class="bold">Save this for your records and DO NOT share it
            with anyone!</p>
            <p class="code">{{createdAccount.getPrivateKeyString()}}</p>
          </div>
        </div>

      </div>

      <div slot="footer">
        <div v-if="!createdAccount">
          <a class="button is-primary" @click="createNewAccount">Create
            New Address</a>
          <a class="button" @click="close">Cancel</a>
        </div>
        <div v-else>
          <a class="button is-primary" @click="close">Close</a>
        </div>
      </div>

    </v-modal>
  </div>
</template>

<script>
import { mapActions } from 'vuex';
import VModal from '@/components/ui/VModal'

export default {
  data () {
    return {
      createdAccount: null
    }
  },
  computed: {
    hdWallet () {
      return this.$store.state.accounts.hdWallet
    },
    // All accounts that have been created
    accounts () {
      return this.$store.state.accounts.accounts
    }
  },
  methods: {
    ...mapActions('accounts', ['addAccount']),
    // Create the next account derived from the HD wallet seed
    // TODO consider gap limit if multiple hd accounts are already used
    createNewAccount() {
      if (!this.hdWallet) {
        return
      }
      let i = this.accounts.length
      let account = this.hdWallet.deriveChild(i).getWallet()
      this.addAccount(account);
      this.createdAccount = account
    },
    close() {
      this.$emit('close')
    }
  },
  components: {
    VModal,
  }
}
</script>

<style lang="scss">
</style>
