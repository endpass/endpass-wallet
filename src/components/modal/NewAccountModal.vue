<template lang="html">
  <div class="new-account-modal">
    <v-modal @close="close">
      <template slot="header">Create New Address</template>

      <div v-if="!isAccountCreated">
        <p class="subtitle">You currently have
        <strong>{{ Object.keys(wallets).length }}</strong> active addresses in your wallet.</p>
        <p class="subtitle">Click the button below to create or import an additional address you can use to receive Ethereum and tokens.</p>
      </div>
      <div v-else>
        <p class="subtitle">New Address Created</p>

        <div class="message">
          <div class="message-header">
            <p>Public Address</p>
          </div>
          <div class="message-body">
            <p>Use this address to receive Ether and tokens.</p>
            <p class="code address">{{ address }}</p>
          </div>
        </div>

        <div class="message is-warning">
          <div class="message-header">
            <p>Private Key</p>
          </div>
          <div class="message-body">
            <p class="bold">Save this for your records and DO NOT share it
            with anyone!</p>
            <p class="code">{{ privateKey }}</p>
          </div>
        </div>

      </div>

      <div
        v-if="!isAccountCreated"
        slot="footer"
        class="buttons"
      >
        <v-form @submit="createNewAccount">
          <v-button
            :loading="isCreatingAccount"
            class-name="is-primary is-medium"
          >Create address</v-button>
        </v-form>
        <v-form @submit="importNewAccount">
          <v-button
            class-name="is-primary is-medium"
          >Import address</v-button>
        </v-form>
      </div>
    </v-modal>
    <password-modal
      v-if="isPasswordModal"
      @confirm="confirmPassword"
      @close="togglePasswordModal"
    />
  </div>
</template>

<script>
import { mapActions, mapState, mapGetters } from 'vuex';
import VModal from '@/components/ui/VModal';
import VInput from '@/components/ui/form/VInput.vue';
import VForm from '@/components/ui/form/VForm.vue';
import VButton from '@/components/ui/form/VButton.vue';
import PasswordModal from '@/components/modal/PasswordModal';

export default {
  name: 'NewAccountModal',
  data() {
    return {
      isAccountCreated: null,
      isCreatingAccount: false,
      privateKey: '',
      isPasswordModal: false,
    };
  },
  computed: {
    ...mapState({
      wallets: state => state.accounts.wallets,
      address: state => state.accounts.address,
    }),
    ...mapGetters('accounts', ['wallet']),
  },
  methods: {
    ...mapActions('accounts', ['generateWallet', 'validatePassword']),
    importNewAccount() {
      this.$router.push('import');
      this.close();
    },
    // Create the next account derived from the HD wallet seed
    // TODO consider gap limit if multiple hd accounts are already used
    createNewAccount() {
      this.togglePasswordModal();
      this.$ga.event({
        eventCategory: 'onboarding',
        eventAction: 'create_new_account',
      });
    },
    togglePasswordModal() {
      this.isPasswordModal = !this.isPasswordModal;
    },
    async confirmPassword(password) {
      this.togglePasswordModal();
      this.isCreatingAccount = true;

      await new Promise(res => setTimeout(res, 20));

      try {
        await this.generateWallet(password);
        this.privateKey = await this.wallet.getPrivateKeyString(password);
        this.isCreatingAccount = false;
        this.isAccountCreated = true;
      } catch (e) {
        this.isCreatingAccount = false;
        this.$notify({
          title: 'Something went wrong',
          text: 'Ð¡ould not create account. Please try again.',
          type: 'is-danger',
        });
      }
    },
    close() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VInput,
    VForm,
    VButton,
    PasswordModal,
  },
};
</script>

<style lang="scss">
</style>
