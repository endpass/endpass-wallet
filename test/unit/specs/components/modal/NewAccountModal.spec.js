import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';

import NewAccountModal from '@/components/modal/NewAccountModal';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('NewAccountModal', () => {
  let wrapper;
  let store;
  let options;

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        accounts: {
          namespaced: true,
          state: {
            address: {
              getChecksumAddressString: () => null,
            },
            wallet: {},
            wallets: {},
          },
          actions: {
            generateWallet: () => {},
            validatePassword: () => {},
          },
        },
      },
    });

    const $ga = { event: jest.fn() };

    options = {
      store,
      localVue,
      mocks: {
        $ga,
      },
    };

    wrapper = mount(NewAccountModal, {
      ...options,
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('NewAccountModal');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should redirect and close the modal for import', () => {
      const spy = jest.spyOn(wrapper.vm, 'close');
      const push = jest.fn();
      wrapper.vm.$router = { push };

      wrapper.vm.importNewAccount();

      expect(spy).toBeCalled();
      expect(push).toBeCalledWith('import');
    });
  });
});
