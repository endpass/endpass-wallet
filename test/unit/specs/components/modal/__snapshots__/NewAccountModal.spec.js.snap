// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAccountModal render should render component 1`] = `
<div
  class="new-account-modal"
>
  <div
    class="modal is-active"
  >
    <div
      class="modal-background"
    />
     
    <div
      class="modal-logo"
    />
     
    <div
      class="modal-card"
    >
      <div
        class="modal-card-head"
      >
        <h3
          class="modal-card-title"
        >
          Create New Address
        </h3>
         
        <button
          class="delete is-large"
        />
      </div>
       
      <div
        class="modal-card-body"
      >
         
        <div>
          <p
            class="subtitle"
          >
            You currently have
      
            <strong>
              0
            </strong>
             active addresses in your wallet.
          </p>
           
          <p
            class="subtitle"
          >
            Click the button below to create or import an additional address you can use to receive Ethereum and tokens.
          </p>
        </div>
         
      </div>
       
      <div
        class="modal-card-foot"
      >
        <div
          class="buttons"
        >
          <form
            class="control"
          >
            <div
              class="field"
            >
              <div
                class="control"
              >
                <button
                  class="button is-primary is-medium"
                >
                  Create address
                </button>
              </div>
            </div>
          </form>
           
          <form
            class="control"
          >
            <div
              class="field"
            >
              <div
                class="control"
              >
                <button
                  class="button is-primary is-medium"
                >
                  Import address
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
   
  <!---->
</div>
`;
